### ISSUE TITLE
Valist publish on non-CI/CD environments requires a mechanism for reproducible build environments

### ISSUE THREAD

**Issue by jiyuu-jin** [valist-io/valist#17]
When running a `valist publish` outside of the valist docker instance there is not currently a streamlined way to ensure that the local build environment is reproducible. 

To enable reproducible environments on local machines valist publish will invoke the docker api to build the users included dockerfile/build script from their local project, then execute the build  of the image and copy the built artifact from the running container to the host machine. 

To perform these actions we have created the `reproducible library` by leveraging the [docker api for node package](https://github.com/apocas/dockerode) Follow the status of the reproducible library at: [Valist Reproducible GitHub](https://github.com/valist-io/reproducible):

An example call to reproducible looks something like this:

```javascript
  await reproducible.runBuild({
      image: 'build-test',
      outputPath: `${process.cwd()}/tests/dist`,
      artifacts: ['main'],
   });
```

**Comment by awantoch** [comment]
some thoughts on the format for the `valist.yml` and how to keep config consistent between desktop/CI
----

in the `valist.yml` file, we'll need the following keys:

```yaml
image: DOCKER_IMAGE
install: INSTALL_COMMAND
build: BUILD_COMMAND
out: OUTPUT_DIR_OR_FILE_INSIDE_CONTAINER
```

the `build` and `install` commands are optional if the `image` that is set already builds what's needed for `out` 

after `image` is set, we run the `install` command, then the `build` command provided inside of a new container (or build a new image from the supplied one). then, we mount and copy the `out` directory on the container that contains the artifact, and copy it to the host + publish.

`image` could also be set to the location of a Dockerfile that's already in the repository, and the rest of the flow stays the same.

eventually, we could/should use a system like buildpacks.io that would auto-detect the package type and start with the relevant container image.

**Comment by jiyuu-jin** [comment]
To that end, maybe something like:

```typescript
const generateDockerfile = (baseImage: string, source: string, buildCommand: string) =>  ` 
  FROM ${baseImage}
  WORKDIR /opt/build
  CP ${source} ./
  WORKDIR /opt/build/${source}
  RUN ${buildCommand}
`;
```

**Comment by jiyuu-jin** [comment]
Oh and +1 to the buildpacks.io, we should add that to the roadmap.

**Comment by awantoch** [comment]
Closed by #28
